parser/src/lexer.rs

```diff
@@ -516,803 +393,533 @@ where
+                    return Ok((Tok::Dedent, TextRange::empty(self.offset())));
                 }
-                Some('\x0C') => {
-                    // Form feed character!
-                    // Reset indentation for the Emacs user.
-                    self.next_char();
-                    spaces = 0;
-                    tabs = 0;
-                }
-                Some('\n' | '\r') => {
-                    // Empty line!
-                    #[cfg(feature = "full-lexer")]
-                    let tok_start = self.get_pos();
-                    self.next_char();
-                    #[cfg(feature = "full-lexer")]
-                    let tok_end = self.get_pos();
-                    #[cfg(feature = "full-lexer")]
-                    self.emit((Tok::NonLogicalNewline, TextRange::new(tok_start, tok_end)));
-                    spaces = 0;
-                    tabs = 0;
-                }
-                None => {
-                    spaces = 0;
-                    tabs = 0;
-                    break;
+            }
+
+            if self.at_begin_of_line && self.nesting == 0 {
+                if let Some(trivia) = self.eat_logical_line_trivia()? {
+                    break Ok(trivia);
                 }
-                _ => {
-                    self.at_begin_of_line = false;
-                    break;
+            }
+
+            self.cursor.start_token();
+            if let Some(c) = self.cursor.bump() {
+                if let Some(normal) = self.consume_normal(c)? {
+                    break Ok(normal);
                 }
+            } else {
+                // Reached the end of the file. Emit a trailing newline token if not at the beginning of a logical line,
+                // empty the dedent stack, and finally, return the EndOfFile token.
+                break self.consume_end();
             }
         }
-
-        Ok(IndentationLevel { tabs, spaces })
     }
 
-    // Push/pop indents/dedents based on the current indentation level.
-    fn handle_indentations(&mut self) -> Result<(), LexicalError> {
-        let indentation_level = self.eat_indentation()?;
+    fn eat_logical_line_trivia(&mut self) -> Result<Option<Spanned>, LexicalError> {
+        let mut indentation = Indentation::root();
+
+        // Eat over any leading whitespace
+        self.cursor.start_token();
+        self.cursor.eat_while(|c| {
+            if c == ' ' {
+                indentation = indentation.add_space();
+                true
+            } else if c == '\t' {
+                indentation = indentation.add_tab();
+                true
+            } else if c == '\x0C' {
+                indentation = Indentation::root();
+                true
+            } else {
+                false
+            }
+        });
 
-        if self.nesting != 0 {
-            return Ok(());
-        }
+        let token = match self.cursor.first() {
+            c @ ('%' | '!' | '?' | '/' | ';' | ',') if self.mode == Mode::Jupyter => {
```

Comment:
Do you want me to merge #30 first or wait until this is merged? As this will have to be moved out from `eat_logical_line_trivia` to emit the `INDENT` token and allow lexing in the assignment statement.

Reply:
Or, I think it would be simpler to move everything from this point onwards until the `lex_magic_command` call i.e., creating the `MagicKind`, into the `lex_magic_command` function itself.

Reply:
You can go ahead with merging. I don't want this to block any other work